File: leetcode/001_two_sum/two_sum.cpp
/*
 * two_sum.cpp
 *
 *  Created on: 2016Äê1ÔÂ27ÈÕ
 *      Author: ncfxy
 */

#include <iostream>
#include <vector>
#include <map>
#include <iterator>
#include <algorithm>
using namespace std;

class Solution {
public:
	vector<int> twoSum(vector<int>& nums, int target) {
		map<int, vector<int> > myMap;
		for (int i = 0; i < nums.size(); i++) {
			if (myMap.find(nums[i]) == myMap.end()) {
				vector<int> ve;
				ve.push_back(i + 1);
				myMap[nums[i]] = ve;
			} else {
				myMap[nums[i]].push_back(i + 1);
			}
		}
		vector<int> ans;
		for (map<int, vector<int> >::iterator it = myMap.begin();
				it != myMap.end(); it++) {
			int temp = target - it->first;
			map<int, vector<int> >::iterator it1 = myMap.find(temp);
			if (it1 != myMap.end()) {
				ans.insert(ans.end(), it->second.begin(), it->second.end());
				if (it != it1) {
					ans.insert(ans.end(), it1->second.begin(),
							it1->second.end());
				}
				break;
			}
		}
		sort(ans.begin(), ans.end());
		return ans;
	}

}
;

int main() {
	cout << "!!!Hello World!!!" << endl; // prints !!!Hello World!!!
	Solution so;
	vector<int> ve;
	ve.push_back(3);
	ve.push_back(5);
	ve.push_back(7);
	ve.push_back(10);
	so.twoSum(ve, 12);
	return 0;
}

File: leetcode/001_two_sum/twoSum.py
class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ans = []
        myMap = {}
        for i in range(len(nums)):
            if nums[i] in myMap:
                myMap[nums[i]].append(i)
            else:
                myMap[nums[i]] = [i]
        
        for key in myMap:
            temp = target - key
            if temp in myMap:
                ans.append(myMap[key][0])
                if(temp == key):
                    ans.append(myMap[key][1])
                else:
                    ans.append(myMap[temp][0])
                break
        ans.sort()
        return ans

so = Solution()
nums = [2, 7, 11, 15]
target = 9
print(so.twoSum(nums, target))
